### DISCUSSION TITLE
Conversation handling using python-bale-bot

### DISCUSSION THREAD

**Discussion by okaeiz** [kianahmadian/python-bale-bot#17] [Q&A] [answered]
In python-telegram-bot, handling conversations can be done through the steps below:
1. We define a function.

```
async def start(update: Update, context: ContextTypes.DEFAULT_TYPE) -> int:
    """Starts the conversation and asks the user about their gender."""
    reply_keyboard = [["Boy", "Girl", "Other"]]

    await update.message.reply_text(
        "Hi! My name is Professor Bot. I will hold a conversation with you. "
        "Send /cancel to stop talking to me.\n\n"
        "Are you a boy or a girl?",
        reply_markup=ReplyKeyboardMarkup(
            reply_keyboard, one_time_keyboard=True, input_field_placeholder="Boy or Girl?"
        ),
    )

    return GENDER
```
2. We define the 'states'.

`GENDER, PHOTO, LOCATION, BIO = range(4)
`

3. Then we create the handler block.
```
def main() -> None:
    """Run the bot."""
    # Create the Application and pass it your bot's token.
    application = Application.builder().token("TOKEN").build()

    # Add conversation handler with the states GENDER, PHOTO, LOCATION and BIO
    conv_handler = ConversationHandler(
        entry_points=[CommandHandler("start", start)],
        states={
            GENDER: [MessageHandler(filters.Regex("^(Boy|Girl|Other)$"), gender)],
            PHOTO: [MessageHandler(filters.PHOTO, photo), CommandHandler("skip", skip_photo)],
            LOCATION: [
                MessageHandler(filters.LOCATION, location),
                CommandHandler("skip", skip_location),
            ],
            BIO: [MessageHandler(filters.TEXT & ~filters.COMMAND, bio)],
        },
        fallbacks=[CommandHandler("cancel", cancel)],
    )

    application.add_handler(conv_handler)

    # Run the bot until the user presses Ctrl-C
    application.run_polling()


if __name__ == "__main__":
    main()
```

Can we do such things here in this library too?

✨The complete reference of the codes above is [here](https://github.com/python-telegram-bot/python-telegram-bot/blob/master/examples/conversationbot.py).

**Comment by kian-ahmadian** [comment, marked as answer]
hey @okaeiz

I did not understand what you meant

Please clearly state what facilities are there in that library that you want to be in python-bale-bot.

**Reply by okaeiz** [reply]
I mean, is there anything like a `ConversationHandler()` using which we can handle different states of conversation?
For example, the bot asks for name, the user sends it, and the data is stored, then, the bot asks for gender, the user chooses it, and the data is stored and so on.

**Reply by kian-ahmadian** [reply]
Of course, as you know, the on_message event is for receiving messages.

Now, we have designed a function for you that you can use to get a message from the user in the middle of the code.

When you use that function, the code will not go to the next line until the user sends that event and the checks you set pass. Also, while using this function, _there is no problem with bot speed in other events and your commands_.

for example:

```python
...
@bot.listen("on_message")
async def when_message(message):
    await message.reply("Please enter your gender (only boy or girl)") 
    gender = await bot.wait_for("message", check = lambda msg: msg.author == message.author and msg.content in ["boy", "girl"] )
    await gender.reply("You're a great {}".format(gender.content))
``` 

You can also set a **timeout** for it:
```python
try:
    await bot.wait_for(..., timeout = 20.0)
except asyncio.TimeoutError:
    print("Timeout!")
else:
    print("The work was done on time")
```

**Reply by okaeiz** [reply]
This `wait_for()` is really useful and I tried it in my code. But what I meant was slightly different. Imagine we have different routes in the conversation. How shall I define these routes?
For example, I have created multiple functions for handling messages and I do not want them to run at the same time. I need something like this: 
```python
"""start command handler"""
@client.listen(EventType.MESSAGE)
async def when_start_command(message: Message):
      #the user will be authenticated here.
      return IS_AUTHENTICATED

"""after the user is authenticated"""
@client.listen(EventType.MESSAGE)
async def when_is_authenticated(message: Message):
     #the user will see the contents here.
```

In other words, I want to able to specify which route I want to take. I have created multiple functions right now but they all run at the same time.

**Reply by kian-ahmadian** [reply]
### I think I understand what you mean

### Depending on your desired performance, you can do the following:

* If you want to check the user before calling all `on_message` events or the like, you can do something similar to #8 .

* But otherwise, you can use the bot.dispatch method in your own way.

You can define events under the command ( **listen("your custom event")** ) and then execute those events by using `bot.dispatch()` in any part of the bot, and giving it your own personalized event.

example:

```python
...
@bot.listen("on_message")
async def when_message(message):
    ...
    await message.reply("Your coffee will be ready soon.")
    bot.dispatch("my_custom_event", message.author)
```

```python
@bot.listen("on_my_custom_event")
async def when_custom_event(client: bale.user):
    ....
```

**Reply by okaeiz** [reply]
So here I use `client.dispatch()` after the users select their gender.

```python
"""start command handler"""
@client.listen(EventType.MESSAGE)
async def when_receive_message(message: Message):
    if message.text in ("/start", "start"):
        await message.reply(text = ("Hi there! To begin using the bot, choose your gender first."),
                            components=Components(
                inline_keyboards=[
                    [
                        InlineKeyboard(
                            "Woman", callback_data="female_client"
                        ),
                        InlineKeyboard("Man", callback_data="male_client"),
                    ],
                ]
            ),
            )
        client.dispatch("GENDER_SELECTED", message.author)
```
And I define the event here:

```python
"""when the users selected their gender"""
@client.listen("GENDER_SELECTED")
async def when_selected_gender(callback: CallbackQuery):
    if callback.data == "male_client":
        await callback.message.reply("Enter your fullname.")
        fullname = await client.wait_for("message")
        welcome_text = f"Welcome, Mr.{fullname.content}"
        await callback.message.reply(welcome_text)
    elif callback.data == "female_client":
        await callback.message.reply("Enter your fullname.")
        fullname = await client.wait_for("message")
        welcome_text = f"Welcome, Mrs.{fullname.content}"
        await callback.message.reply(welcome_text)

```
But it doesn't work right now.

**Reply by kian-ahmadian** [reply]
### Two points are not observed in this code.

* In the **`bot.dispatch`** method, the method you write inside should be like this:

```python
...
@bot.listen("on_message")
async def when_message(message):
    ...
    await message.reply("Your coffee will be ready soon.")
    bot.dispatch("my_custom_event", message.author)
```

```python
@bot.listen("on_my_custom_event")
async def when_custom_event(client: bale.user):
    ....
```

You did not leave the **`on`** statement for the listener.

* Nowhere in the code did you wait for the callback sent by the user. Therefore, you should have used **`bot.wait_for`** before using **`bot.dispatch`.** And then you would send the callback value for that event.

**Reply by okaeiz** [reply]
Ah, quite. Thanks for that.
Therefore, `my_custom_event` which is passed to `bot.dispatch()` is the same as `on_my_custom_event` which is passed to `@bot.listen()`? 
If they are both the event name, why aren't they identical?

**Reply by kian-ahmadian** [reply]
### Yes. That's right.

This is the property of this package to make it easier to write and read the code, and `on` the beginning

**Reply by okaeiz** [reply]
what should be my `wait_for` line before `bot.dispatch()`?
Is it something like this?
```python
await client.wait_for("callback")
```

**Reply by kian-ahmadian** [reply]
```python
...
callback = await client.wait_for("callback", ...)
# it's After get callback and user click on a inline keyboard button
bot.dispatch("send_coffee", callback)
```

ok?
